File: org.eclipse.scout.sdk.s2e/src/main/java/org/eclipse/scout/sdk/s2e/util/ApiHelper.java
/*
 * Copyright (c) 2010, 2023 BSI Business Systems Integration AG
 *
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 */
package org.eclipse.scout.sdk.s2e.util;

import static org.eclipse.scout.sdk.s2e.environment.EclipseEnvironment.callInEclipseEnvironmentSync;

import java.util.Optional;
import java.util.function.Function;

import org.eclipse.core.runtime.NullProgressMonitor;
import org.eclipse.jdt.core.IJavaElement;
import org.eclipse.jdt.core.IJavaProject;
import org.eclipse.scout.sdk.core.java.apidef.ApiVersion;
import org.eclipse.scout.sdk.core.java.apidef.IApiSpecification;
import org.eclipse.scout.sdk.core.s.java.apidef.IScoutApi;
import org.eclipse.scout.sdk.core.s.java.apidef.ScoutApi;
import org.eclipse.scout.sdk.s2e.environment.EclipseEnvironment;

public final class ApiHelper {

  private ApiHelper() {
  }

  public static IScoutApi requireScoutApiFor(IJavaElement context) {
    return requireScoutApiFor(context, null);
  }

  public static IScoutApi requireScoutApiFor(IJavaElement context, EclipseEnvironment environment) {
    var project = context.getJavaProject();
    return applyWithEclipseEnvironment(e -> requireApiFor(project, IScoutApi.class, e), environment);
  }

  public static Optional<ApiVersion> scoutVersionOf(IJavaElement context, EclipseEnvironment environment) {
    var project = context.getJavaProject();
    return applyWithEclipseEnvironment(e -> ScoutApi.version(e.toScoutJavaEnvironment(project)), environment);
  }

  static <T> T applyWithEclipseEnvironment(Function<EclipseEnvironment, T> task, EclipseEnvironment existingEnvironment) {
    if (existingEnvironment == null) {
      return callInEclipseEnvironmentSync((e, p) -> task.apply(e), new NullProgressMonitor());
    }
    return task.apply(existingEnvironment);
  }

  public static <T extends IApiSpecification> T requireApiFor(IJavaProject project, Class<T> api, EclipseEnvironment environment) {
    return applyWithEclipseEnvironment(e -> e.toScoutJavaEnvironment(project).requireApi(api), environment);
  }

  public static Optional<IScoutApi> scoutApiFor(IJavaProject project) {
    return scoutApiFor(project, null);
  }

  public static Optional<IScoutApi> scoutApiFor(IJavaProject project, EclipseEnvironment environment) {
    return apiFor(project, IScoutApi.class, environment);
  }

  public static <T extends IApiSpecification> Optional<T> apiFor(IJavaProject project, Class<T> api) {
    return apiFor(project, api, null);
  }

  public static <T extends IApiSpecification> Optional<T> apiFor(IJavaProject project, Class<T> api, EclipseEnvironment environment) {
    return applyWithEclipseEnvironment(e -> e.toScoutJavaEnvironment(project).api(api), environment);
  }
}

File: org.eclipse.scout.sdk.s2i/src/main/kotlin/org/eclipse/scout/sdk/s2i/util/ApiHelper.kt
/*
 * Copyright (c) 2010, 2023 BSI Business Systems Integration AG
 *
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 */
package org.eclipse.scout.sdk.s2i.util

import com.intellij.openapi.module.Module
import com.intellij.psi.PsiElement
import org.eclipse.scout.sdk.core.apidef.IApiSpecification
import org.eclipse.scout.sdk.core.s.apidef.IScoutApi
import org.eclipse.scout.sdk.core.util.Ensure.newFail
import org.eclipse.scout.sdk.s2i.containingModule
import org.eclipse.scout.sdk.s2i.environment.IdeaEnvironment
import org.eclipse.scout.sdk.s2i.environment.IdeaEnvironment.Factory.callInIdeaEnvironmentSync
import org.eclipse.scout.sdk.s2i.environment.IdeaProgress

object ApiHelper {

    fun requireScoutApiFor(context: PsiElement, environment: IdeaEnvironment? = null) = scoutApiFor(context, environment)
            ?: throw newFail("Cannot find a module for context element: {}", context)

    fun scoutApiFor(context: PsiElement, environment: IdeaEnvironment? = null) = context.containingModule()
            ?.let { scoutApiFor(it, environment) }

    fun scoutApiFor(module: Module, environment: IdeaEnvironment? = null): IScoutApi? {
        return apiFor(module, IScoutApi::class.java, environment)
    }

    fun <T : IApiSpecification> apiFor(module: Module, api: Class<T>, environment: IdeaEnvironment? = null): T? {
        if (environment == null) {
            return callInIdeaEnvironmentSync(module.project, IdeaProgress.empty()) { env, _ ->
                createApiFor(module, api, env)
            }
        }
        return createApiFor(module, api, environment)
    }

    private fun <T : IApiSpecification> createApiFor(module: Module, api: Class<T>, environment: IdeaEnvironment): T? {
        val env = environment.toScoutJavaEnvironment(module) ?: throw newFail("Module '{}' is no Java module.", module.name)
        return env.api(api).orElse(null)
    }
}